src/pool.rs

```diff
@@ -219,28 +218,32 @@ impl PoolKey {
     }
 }
 
-/// Read wrapper that returns the stream to the pool once the
+/// Read wrapper that returns a stream to the pool once the
 /// read is exhausted (reached a 0).
 ///
 /// *Internal API*
 pub(crate) struct PoolReturnRead<R: Read + Sized + Into<Stream>> {
-    // unit that contains the agent where we want to return the reader.
-    unit: Option<Box<Unit>>,
-    // wrapped reader around the same stream
+    // the agent where we want to return the stream.
+    agent: Agent,
+    // wrapped reader around the same stream. It's an Option because we `take()` it
+    // upon returning the stream to the Agent.
     reader: Option<R>,
+    // Key under which to store the stream when we're done.
+    key: PoolKey,
 }
 
 impl<R: Read + Sized + Into<Stream>> PoolReturnRead<R> {
-    pub fn new(unit: Option<Box<Unit>>, reader: R) -> Self {
+    pub fn new(agent: &Agent, url: &Url, reader: R) -> Self {
         PoolReturnRead {
-            unit,
+            agent: agent.clone(),
+            key: PoolKey::new(url, agent.config.proxy.clone()),
             reader: Some(reader),
         }
     }
 
     fn return_connection(&mut self) -> io::Result<()> {
         // guard we only do this once.
-        if let (Some(unit), Some(reader)) = (self.unit.take(), self.reader.take()) {
+        if let Some(reader) = self.reader.take() {
```

Comment:
It's still true - we can only return the stream once because we `.take()` it. The comment is a little funny because we don't really need to take special steps to guard it - Rust's type system won't let us do this more than once, because after `.take()`ing the reader once, we can't possibly take it a second time. I think it's okay as-is but if there's something you think it should say instead (or if we should delete it) lemme know!